test/distributed/_sharded_tensor/test_megatron_prototype.py

```diff
@@ -0,0 +1,222 @@
+# Owner(s): ["oncall: distributed"]
+
+import copy
+import sys
+
+import torch
+import torch.distributed as dist
+from torch.distributed._sharded_optim import (
+    ShardedOptimizer,
+    named_params_with_sharded_tensor,
+)
+from torch.distributed._sharded_tensor import (
+    shard_parameter,
+    _collect_local_shard,
+    _reshard_output,
+)
+from torch.testing._internal.common_distributed import (
+    requires_nccl,
+    skip_if_lt_x_gpu,
+)
+from torch.testing._internal.common_utils import (
+    TEST_WITH_DEV_DBG_ASAN,
+    run_tests,
+)
+from torch.testing._internal.distributed._sharded_tensor import (
+    TEST_GPU_NUM,
+    ShardedTensorTestBase,
+    with_comms,
+)
+from torch.testing._internal.distributed._sharded_tensor._test_ops_common import (
+    clone_module_parameter,
+    generate_chunk_sharding_specs_for_test,
+    generate_local_weight_sharding_params_for_test,
+)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class TestShardedTensorMegatronLinear(ShardedTensorTestBase):
+    class SimpleMegatronLM(torch.nn.Module):
+        def __init__(self, linear_size, rank=None):
+            super().__init__()
+            self.fc1 = torch.nn.Linear(*linear_size[0])
+            self.gelu = torch.nn.GELU()
+            self.fc2 = torch.nn.Linear(*linear_size[1])
+            if rank:
+                self.fc1.cuda(rank)
+                self.fc2.cuda(rank)
+
+        def forward(self, inp):
+            return self.fc2(self.gelu(self.fc1(inp)))
+
+    def _run_megatron_linear(self, spec, input_size, linear_size):
+        def _weight_override(module_dst, module_src):
+            module_dst.fc1.weight = clone_module_parameter(module_src.fc1, "weight")
+            module_dst.fc1.bias = clone_module_parameter(module_src.fc1, "bias")
+            module_dst.fc2.weight = clone_module_parameter(module_src.fc2, "weight")
+            module_dst.fc2.bias = clone_module_parameter(module_src.fc2, "bias")
+
+        def _shard_parameter(module, spec):
+            shard_parameter(module.fc1, "weight", spec[0])
+            shard_parameter(module.fc2, "weight", spec[1])
+
+        def _get_weight_grad(module):
+            return (module.fc1.weight.grad, module.fc2.weight.grad)
+
+        def _get_bias_grad(module):
+            return (module.fc1.bias.grad, module.fc2.bias.grad)
+
+        def _get_weights(module):
+            return (module.fc1.weight, module.fc2.weight)
+
+        def _get_bias(module):
+            return (module.fc1.bias, module.fc2.bias)
+
+        def _get_weight_local_shard(module):
+            return (
+                module.fc1.weight.local_shards()[0].tensor,
```

Comment:
nit: You can use the new `local_shard()` api :)

Reply:
Sure.